12-offer-encoding.md

```diff
@@ -391,65 +407,88 @@ for [Signature Calculation](#signature-calculation).
+    - if `offer_amount` is present:
+      - MUST specify `invoice_request_amount`.
+    - otherwise:
+      - MAY omit `invoice_request_amount`.
+      - if it sets `invoice_request_amount`:
+        - MUST specify `invoice_request_amount`.`msat` as greater or equal to amount expected by `offer_amount` (and, if present, `offer_currency`).
+    - MUST set `invoice_request_payer_id` to a transient public key.
+    - MUST remember the secret key corresponding to `invoice_request_payer_id`.
+  - otherwise (not responding to an offer):
+    - MUST set (or not set) `offer_metadata`, `offer_description`, `offer_absolute_expiry`, `offer_paths` and `offer_issuer` as it would for an offer.
+    - MUST NOT include `signature`, `offer_chains`, `offer_amount`, `offer_currency`, `offer_features`, `offer_quantity_min`, `offer_quantity_max` or `offer_node_id`
+    - if it sets `invoice_request_payer_id`:
+      - MUST set `key` to the node's public key to request the invoice from.
+    - otherwise:
+       - MUST provide at least one `offer_blinded_path`
+    - if the chain for the invoice is not solely bitcoin:
+      - MUST specify `invoice_request_chain` the offer is valid for.
+    - MUST set `invoice_request_amount`.
   - MUST NOT set any tlv fields greater or equal to 160.
   - SHOULD set `invoice_request_payer_info` to an unpredictable series of bytes.
-  - MUST set `invoice_request_payer_key` to a transient public key.
-  - MUST remember the secret key corresponding to `invoice_request_payer_key`.
-  - MUST NOT set or imply any `invoice_request_chain` not set or implied by `offer_chains`.
-  - MUST set `signature`.`sig` as detailed in [Signature Calculation](#signature-calculation) using the `invoice_request_payer_key`.
   - if `offer_quantity_min` or `offer_quantity_max` are present:
     - MUST set `invoice_request_quantity`
     - MUST set it within that (inclusive) range.
   - otherwise:
     - MUST NOT set `invoice_request_quantity`
-  - if `offer_amount`:
-    - MUST specify `invoice_request_amount`.`msat` in multiples of the minimum lightning-payable unit
-      (e.g. milli-satoshis for bitcoin) for `chain` (or for bitcoin, if there is no `chain`).
-  - otherwise:
-    - MAY omit `invoice_request_amount`.
-    - if it sets `invoice_request_amount`:
-      - MUST specify `invoice_request_amount`.`msat` as greater or equal to amount expected by `offer_amount` (and, if present, `offer_currency`).
+  - if it sets `invoice_request_amount`:
+    - MUST set `msat` in multiples of the minimum lightning-payable unit
+        (e.g. milli-satoshis for bitcoin) for `invoice_request_chain` (or for bitcoin, if there is no `invoice_request_chain`).
   - if it supports bolt12 features:
     - MUST set `invoice_request_features`.`features` to the bitmap of features.
 
 The reader:
-  - MUST fail the request if `invoice_request_payer_key` is not present.
+  - MUST fail the request if `invoice_request_payer_id` is not present.
```

Comment:
@rustyrussell This shouldn't hold if the invoice is in response to an `inv_req` QR, right? I believe the `invoice_request_payer_id` is optional in that case since there's no signature? 

@jkczyz if this is correct, then I think the resolution of https://github.com/rustyrussell/lightning-rfc/pull/11#issuecomment-1226526354 is that we'll never set these fields for QRs and always just set `offer_metadata`, so `invoice_request` ends up being a bit smaller than `offer_send_invoice`s but comparable

Reply:
I ended up dropping the whole "node_id is optional if you use a blinded path and don't have a real node id" since it just made this complicated.  So that would apply here too, and it might be important later if you want to later prove your id?  It's certainly harmless to include one.

Reply:
Hmmm, aren't we trying to save bits on the QR code here? If I'm missing something, though, I think the requirements for the `writer` need updating for this in the QR case

Reply:
Once you have blinded paths in the QR code, it's hard to argue that saving 35 bytes is a big win.  It's possible to allow this case later, but for now we have more than enough complexity :(

Reply:
35 bytes is pretty significant in a QR code, plus even a single 2-hop blinded path can add a lot of privacy so I'm not convinced that "blinded paths == we should give up entirely on the QR case" IMHO.

Not gonna die on this hill but wanted to put out those points :p it's true we *can* add it later as you say

Reply:
FWIW, it's actually 56 bytes after bech32 encoding, IIUC.